ibis/util.py

```diff
@@ -407,3 +407,56 @@ def wrapper(*args, **kwargs):
         return wrapper
 
     return decorator
+
+
+class Eq:
+    """A mixin for abstracting equality checks."""
+
+    __slots__ = ()
+
+    def __eq__(self, other: Any) -> bool:
+        return self.equals(other, cache={})
+
+    def __ne__(self, other: Any) -> bool:
+        return not (self == other)
+
+    def equals(
+        self,
+        other: Any,
+        cache: MutableMapping[Hashable, bool] | None = None,
+    ) -> bool:
+        return self is other or (
+            type(self) == type(other) and self._raw_equals(other, cache=cache)
+        )
+
+    def _raw_equals(
```

Comment:
Perhaps `__equals__` or that would be too confusing?

Reply:
How about `__component_eq__`?

Reply:
Actually __field_eq__ used at the datatypes looks like a better name. It should make the port of datatypes to be annotable easier too.